### PR TITLE
add a random access ledger format

### PR DESCRIPTION
**Pull request by rob-solana**
next steps:

1. move write_stage to use LedgerWriter
2. move fullnode to LedgerReader
   2a. change bin/fullnode to take a ledger directory
3. move multinode tests to allow validators to have their own ledger directories
4. add LedgerWriter to validators (somewhere)
5. add LedgerWindow to run_window_request
6. ??

### CODE REVIEW

File: src/ledger.rs [hunk 1]
```diff
@@ -1,18 +1,162 @@
...
+    pub fn new(directory: &str) -> io::Result<Self> {
+        let directory = Path::new(&directory);
+
+        let index = File::open(directory.join("index"))?;
+        let data = File::open(directory.join("data"))?;
+
+        Ok(LedgerWindow { index, data })
+    }
+
+    pub fn get_entry(&mut self, index: u64) -> io::Result<Entry> {
```
Comment (garious) [inline src/ledger.rs]: Implement the Index trait instead? https://doc.rust-lang.org/std/ops/trait.Index.html

File: src/ledger.rs [hunk 2]
```diff
@@ -1,18 +1,162 @@
...
+
+#[derive(Debug)]
+pub struct LedgerReader {
+    offset: u64, // next start_offset
+    index: File,
+    data: File,
+}
+
+impl Iterator for LedgerReader {
+    type Item = io::Result<Entry>;
```
Comment (garious) [inline src/ledger.rs]: How about a `bincode::Result`?

File: src/ledger.rs (hunk 1 above)
Comment (rob-solana) [reply src/ledger.rs]: Index takes &self, I need &mut self...

File: src/ledger.rs (hunk 2 above)
Comment (rob-solana) [reply src/ledger.rs]: will raise an issue for this one, something like "give ledger its own errors" or "look into error_chain across the board for solana", or "make ledger able to return multiple types of errors" ?

**Pull request merged**